### PR TITLE
[Sema] Only emit warnings on entry point attrs with missing shader attr for libraries

### LINKED ISSUE #6100: Add warning for entry-only attribute without the shader attribute when compiling to library target
**Is your feature request related to a problem? Please describe.**

When Compiling to a library target, such as `lib_6_7`, DXC used to infer that a function was an entry point of a specific kind if it used an attribute that was only meaningful on an entry point of a particular shader kind.  This was broken for at least compute vs. amplification vs. mesh, since it assumed that the `[numthreads(...)]` attribute implied compute shader, even though it is now used with other shader kinds as well.

This shader kind inference was not meant to be used to indicate entry points in libraries, since the HLSLShaderAttribute, or `[shader("<kind>")]` in HLSL, was introduced for this purpose.  This inference was originally meant for determining whether you had attributes indicating a conflicting shader kind when compiling an entry point to a certain shader target.

These facts led us to remove this shader entry point inference for libraries, since it can cause unintended entry points to be included in your library, and it can make an incorrect assumption in ambiguous cases (ex. numthreads).

For shader kinds that are supported in libraries by the D3D runtime (DXR entry points, and node shaders in the future), you have always been required to use the shader attribute (`[shader("<kind>")]`) to identify a function as an entry point.  So that change did not cause a problem for library compilations used for DXR.

However, there may be users of offline linking that relied on this entry point inference in intermediate libraries used for linking to different final shader targets.  In these scenarios that used to rely on this inference, the entry point will now be seen as an ordinary internal function, which will be eliminated as unused in the output when compiling to a library target.  This would cause subsequent linking to fail.  The fix is to add the `[entry("kind")]` attribute to the function, but the root cause and this fix may not be obvious when upgrading to the latest compiler.

**Describe the solution you'd like**
Add a warning when compiling to a library target (`lib_6_*`), when a function lacks the shader attribute (`[shader("<kind>")]`), and includes any other attributes which are only meaningful when applied to an entry point.

Example attributes used to indicate an entry point function: `numthreads`, `rootsignature`, `clipplanes`, `outputcontrolpoints`, and many more.

Some of these attributes are only used with a specific shader kind (like `clipplanes` indicating "vertex" shader), or a limited subset of shader kinds (like `numthreads` indicating "compute", "amplification", "mesh", or "node").  It might be useful to identify the legal kinds usable with the attribute, or the intersecting set from such attributes found, and use this in a suggestion to add the `[shader("<kind>")]` attribute with the identified shader kinds that are compatible with the other attributes found.

It's important that this warning is assigned a group to allow it to be turned off, so that intentional cases that cause the warning to fire can be muted when treating warnings as errors (-WX).

**Describe alternatives you've considered**
- Rely only on a release note about the change in entry point inference for library compilations.
  - People often don't read or remember all the release notes or may not really comprehend their full impact when reading them.  Also, the people who run into the problem may not have been the ones upgrading the compiler and reading those notes.  It's best to emit an explicit warning with a recommended fix when we detect a case where we may not be interpreting things the way the developer intended.
- Make this an error instead of a warning
  - This is problematic because it's legal to have shader entry points in HLSL that you don't intend to use in a library compilation, or for these attributes to be used and ignored on non-entry functions for a given compilation.  For instance, you might call the function from an entry point in your library, and for another compilation you use the same function as the direct entry point.
  - Options can be used to turn all warnings, or just this warning, into an error as desired, so this seems unnecessary.

### PR DESCRIPTION
**Pull request by bob80905**
In implementing this PR:
https://github.com/microsoft/DirectXShaderCompiler/pull/6243
somehow, the condition to only emit the warning for library shaders was left out.
This PR is a follow up on #6243 to properly restrict the emission of this warning for non-library shaders.
This should properly fix #6100

### CODE REVIEW

**Review by python3kgae — APPROVED** [review]

File: tools/clang/include/clang/Basic/DiagnosticGroups.td [hunk 1]
```diff
@@ -783,6 +783,7 @@ def HLSLEffectsSyntax : DiagGroup<"effects-syntax">;
 def HLSLForRedefinition : DiagGroup<"for-redefinition">;
 def HLSLAttributeStatement : DiagGroup<"attribute-statement">;
 def HLSLAttributeType : DiagGroup<"attribute-type">;
+def HLSLEntryAttributeWithoutShaderAttrType : DiagGroup<"entry-attribute-type-without-shader-attribute-type">;
```
Comment (pow2clk) [inline tools/clang/include/clang/Basic/DiagnosticGroups.td:786]: It's kinda long. Maybe something as simple as "non-entry-attributes", "misplaced-attributes"?

**Review by pow2clk — APPROVED** [review]
I think the group name is a bit unwieldy. Otherwise this is good.

**Review by pow2clk — APPROVED** [review]
Looks good!

**Comment by tex3d** [comment]
Sorry I missed the review of this earlier.  While `misplaced-attributes` is an improvement, I think something like `ignored-entry-attributes` might have been more accurate/appropriate.

I also realize that the warning could have been worded a little more clearly for the impact.  It's currently "attribute '%0' ignored without accompanying shader attribute", but I think this would have been better "entry attribute %0 ignored; entry functions require [shader(...)] attribute for inclusion in library".

**Pull request merged**